compiler/rustc_typeck/src/check/method/confirm.rs

```diff
@@ -176,15 +188,32 @@ impl<'a, 'tcx> ConfirmContext<'a, 'tcx> {
                 adjustments.push(Adjustment { kind: Adjust::Pointer(PointerCast::Unsize), target });
             }
         } else {
-            // No unsizing should be performed without autoref (at
-            // least during method dispach). This is because we
-            // currently only unsize `[T;N]` to `[T]`, and naturally
-            // that must occur being a reference.
+            // No unsizing should be performed without autoref (at least during
+            // method dispach). This is because we currently only unsize `[T;N]`
+            // to `[T]`, and naturally that must occur being a reference.
             assert!(pick.unsize.is_none());
+
+            if pick.to_const_ptr {
+                target = match target.kind() {
+                    ty::RawPtr(ty::TypeAndMut { ty, mutbl }) => {
+                        assert_eq!(*mutbl, hir::Mutability::Mut);
+                        self.tcx.mk_ptr(ty::TypeAndMut { mutbl: hir::Mutability::Not, ty })
+                    }
+                    _ => panic!(),
```

Comment:
Also, why is panic a reasonable thing to do here?

Reply:
If `pick.to_const_ptr` is set then the receiver must be a `*mut` and the method's self type must be `*const`. `to_const_ptr` basically says "coerce mut ptr to const ptr". So if it's set, then target must be `RawPtr`, otherwise it's a bug and we shouldn't have set `to_const_ptr = true`.

I will add a panic message.